spec/integration/rails/finders_spec.rb

```diff
@@ -1095,6 +1099,71 @@ def show
             .to eq([other_table_hobby1.id, other_table_hobby2.id])
         end
       end
+
+      context 'when polymorphic many-to-many' do
+        let(:tag1) { Legacy::Tag.create!(name: 'One') }
+        let(:tag2) { Legacy::Tag.create!(name: 'Two') }
+
+        context 'when authors' do
+          before do
+            author1.tags << tag1
+            author2.tags << tag1
+            author2.tags << tag2
+          end
+
+          it 'still works' do
+            do_index({ include: 'tags' })
+            sl = d[0].sideload(:tags)
+            expect(sl.map(&:id)).to eq([tag1.id])
+            sl = d[1].sideload(:tags)
+            expect(sl.map(&:id)).to eq([tag1.id, tag2.id])
+          end
+        end
+
+        context 'when books' do
+          before do
+            book1.tags << tag1
+            book1.tags << tag2
+            book2.tags << tag1
+          end
+
+          it 'still works' do
+            allow(Legacy::BookResource).to receive(:validate_endpoints?) { false }
+            allow(controller).to receive(:resource) { Legacy::BookResource }
+            do_index({ include: 'tags' })
+            sl = d[0].sideload(:tags)
+            expect(sl.map(&:id)).to eq([tag1.id, tag2.id])
+            sl = d[1].sideload(:tags)
+            expect(sl.map(&:id)).to eq([tag1.id])
+          end
+        end
+
+        if ::Rails::VERSION::MAJOR >= 5
+          context 'when both' do
+            let(:tag3) { Legacy::Tag.create!(name: 'Three') }
+
+            before do
+              author1.tags << tag1
+              book2.tags << tag2
+              book2.tags << tag3
+            end
+
+            it 'still works' do
+              allow(Legacy::TagResource).to receive(:validate_endpoints?) { false }
+              allow(controller).to receive(:resource) { Legacy::TagResource }
+              do_index({
+                filter: {
+                  taggable_id: [
+                    { id: author1.id, type: author1.class.name }.to_json,
```

Comment:
If I'm reading this correctly, the query here would look something like:

```
{ id: 12345, type: 'Legacy::Author' }
```

This is incorrect, IMO.  Instead we should be expecting a jsonapi type and id pair, as the current state exposes too much internal implementation detail.

Reply:
It's an interesting point, but not sure if I agree.

Let's say we had a random STI table - no relationships. The ED sample app has `Task` for instance which has a type column with values `Bug`, `Epic`, etc. If you were to query this API, would you expect

`?filter[type]=Bug`

or

`?filter[type]=bugs`

I'd say the former - the `type` here is acting like any other attribute.

So I guess the question here is, since this is a relationship and the JSON is in a particular format, should we treat this specially? I think you can make a case for that, and if anyone wanted to PR with a configuration option I'd support it. But to me the "principle of least surprise", that is most closely following existing usage, is to pass the value explicitly instead of adding this extra indirection.

Reply:
Yeah I can get behind that from a "what does this do by default?" perspective. I think I probably wouldn't have even thought to say something here if we weren't happening to use a namespaced class in our test code, which reveals more explicitly that we're exposing internal implementation to our API.  

Maybe we merge this, but add a TODO to either the core gem or our cookbooks for a pattern for how to make _both_ of these APIs hide those internal Rails STI/polymorphic associations things away from the outside world.  That way either here or in the simple STI case, the `type` field would be some API-level abstraction and not the AR model we happen to be dealing with.  Not sure where that code _should_ live, in the end.

Reply:
Same page. I think a cookbook would be good for this - even if it's not high-usage, it's a good way to show how to customize things. I've added to my notes, let's get a cookbook section rolling out soon.